### ISSUE TITLE
wrapping op in while_loop makes it run faster, even with single iteration

### ISSUE THREAD

**Issue by apisutilis** [tensorflow/tensorflow#16942]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: v1.5.0-0-g37aa430 (1.5.0)
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: 0.9.0
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: CUDA 9.1, CuDNN 7
- **GPU model and memory**: GeForce GTX 1050, 4042MiB
- **Exact command to reproduce**: See below

### Describe the problem

I am working on an object detector. I have one entry point to my code that detects objects in a single image, and another that reads in a batch of images. The batch version uses `tf.while_loop` to apply the same inference code to each image in the batch. I can't put all the images into a single tensor because they are all likely to be different dimensions, so I am using a `TensorArray`. The final results work as expected, but the timing behavior is odd.

The first `sess.run` call in the loop over the dataset is always slower than the rest, presumably due to caching, memory allocation, etc. However, the first call in the non-batch version is WAY slower than the first call in the batch version. I distilled this down to the minimal examples below. The first version does a dot product in a straightforward way. On my machine, it takes about 13 seconds to run. The second version does the same dot product, but wrapped in a `while_loop` and writing the result to an intermediate `TensorArray` and then reading it out again. On my machine, it takes about 4 seconds to run. Presumably, version 2 has to do the same data copying, memory allocation, etc. as version 1, in addition to the overhead of the `while_loop` and `TensorArray` calls. So, the timing behavior is unexpected.

Version 2 can even be modified to compute the same dot product 10 (or more) times, and it will still run faster. This behavior holds even if the loop body is modified to use a random vector every time, so I don't think it is due to caching. It also holds even if `parallel_iterations` is held to `1`, so I don't think it is due to parallelism. Is this a bug, or is something else going on under the hood?

### Source code / logs

```python
# version 1 (~13 seconds on my machine)
graph = tf.Graph()
with graph.as_default(), tf.device('/gpu:0'):
    nums = tf.range(200000000, dtype=tf.float32)
    dot_product = tf.reduce_sum((nums/2) * (nums-5))
with tf.Session(graph=graph) as sess:
    print(sess.run(dot_product))
```

```python
# version 2 (~4 seconds on my machine)
graph = tf.Graph()
with graph.as_default(), tf.device('/gpu:0'):
    arr = tf.TensorArray(size=1, dtype=tf.float32)
    nums = tf.range(200000000, dtype=tf.float32)
    i = tf.constant(0)
    def body(i, arr):
        arr = arr.write(i, tf.reduce_sum((nums/2) * (nums-5)))
        return i+1, arr
    i, arr = tf.while_loop(
        cond=lambda i, x: i < 1,
        body=body,
        loop_vars=[i, arr],
        parallel_iterations=1)
    dot_product = arr.read(0)
with tf.Session(graph=graph) as sess:
    print(sess.run(dot_product))
```

**Comment by Joshuaalbert** [comment]
I find the same behaviour.
I include a slight modified copy-pastable MVCE.
I get 9s and 3 s with the below on CPU.

## MVCE
```
import tensorflow as tf
from timeit import default_timer

def func(nums):
    return tf.reduce_sum(tf.multiply(tf.add(nums,1), nums))

def v1():
    graph = tf.Graph()
    with graph.as_default(), tf.device('/cpu:0'):
        nums = tf.range(200000000, dtype=tf.float32)
        dot_product = func(nums)
    with tf.Session(graph=graph) as sess:
        print(sess.run(dot_product))

def v2():
    graph = tf.Graph()
    with graph.as_default(), tf.device('/cpu:0'):
        arr = tf.TensorArray(size=1, dtype=tf.float32)
        nums = tf.range(200000000, dtype=tf.float32)
        i = tf.constant(0)
        def body(i, arr):
            arr = arr.write(i, func(nums))
            return i+1, arr
        i, arr = tf.while_loop(
            cond=lambda i, x: i < 1,
            body=body,
            loop_vars=[i, arr],
            parallel_iterations=1)
        dot_product = arr.read(0)
    with tf.Session(graph=graph) as sess:
        print(sess.run(dot_product))

if __name__ == '__main__':
    t1 = default_timer()
    v1()
    print("time v1: {}".format(default_timer() - t1))
    t1 = default_timer()
    v2()
    print("time v2: {}".format(default_timer() - t1))
```

**Comment by Joshuaalbert** [comment]
I'll try wrapping in a profiler context to see if I can spot the difference.

**Comment by prb12** [comment]
Seems strange - since v2 should be a superset of the ops in v1.  Perhaps a difference in something like constant folding?   (maybe try changing the optimization level in the session options) 

Probably easiest to just grab a timeline of each (or use the profiler) and see which ops actually get run.

Also, several of the tensors involved are 1.6GB so this *might* cause some interesting differences in memory allocation?

**Comment by prb12** [comment]
The first session.run call being slower is almost certainly due to running the constant folder to optimize the graph.  When session.run is called for the first time on a set of feeds/fetches, the constant folder will run and parts of the graph will be replaced with a tf.constant node that has a protobuf containing the constant values.  Potential constants are always evaluated on the CPU device.  (Also, if the resulting value is larger than a certain threshold then it is discarded and the graph is left alone).

It it likely that when the same expression is computed in a loop the constant folder doesn't consider it a candidate.

In your code snippets above, you only time a single session.run call for each variant...   I would expect constant folding overheads (and many other overheads) to only affect the first execution of each graph.  

If you do a warmup sess.run(dot_product) call, then put a loop around multiple calls of sess.run(dot_product) and evaluate it repeatedly, do you still see the loop version runnig faster?   If so then this is worth looking into further!

i.e. something like:
```
with tf.Session(graph=graph) as sess:
   print(sess.run(dot_product))
   t1 = default_timer()
   for _ in xrange(10):
       sess.run(dot_product)
   print("time v1: {}".format(default_timer() - t1))
```

**Comment by Joshuaalbert** [comment]
So I ran sess.run first and then 10 runs in a timing loop with the constant variable, and also with a random variable. I also created a profile timeline for each.

Constant:
V1 timing: 0.00014590219943784178 / run 10 runs
V2 timing: 0.2790762079996057 / run 10 runs

Random:
V1 timing: 1.2271329755021725 / run 10 runs
V2 timing: 1.209720030199969 / run 10 runs

So it seems while loop is slower as expected after initial graph run.

``` python
import tensorflow as tf
from timeit import default_timer

def func(nums):
    return tf.reduce_sum(tf.multiply(tf.add(nums,1), nums))

def v1(random=False):
    graph = tf.Graph()
    with graph.as_default(), tf.device('/cpu:0'):
        if random:
            nums = tf.random_normal(shape=(200000000,),dtype=tf.float32)
        else:
            nums = tf.range(200000000, dtype=tf.float32)
        dot_product = func(nums)
    with tf.Session(graph=graph) as sess:
        sess.run(dot_product)
        t1 = default_timer()
        for i in range(10):
            sess.run(dot_product)
        print("V1 timing: {} / run 10 runs".format((default_timer() - t1)/10.))

        profiler = tf.profiler.Profiler(sess.graph)
        run_meta = tf.RunMetadata()
        sess.run(dot_product,
                options = tf.RunOptions(trace_level = tf.RunOptions.FULL_TRACE),
                run_metadata = run_meta)
        profiler.add_step(0,run_meta)
        opts = (tf.profiler.ProfileOptionBuilder(
                 tf.profiler.ProfileOptionBuilder.time_and_memory())
                 .with_step(0)
                 .with_timeline_output('timeline_v1_{}'.format('random' if random else 'const')).build())
        profiler.profile_graph(options=opts)



def v2(random=False):
    graph = tf.Graph()
    with graph.as_default(), tf.device('/cpu:0'):
        arr = tf.TensorArray(size=1, dtype=tf.float32)
        if random:
            nums = tf.random_normal(shape=(200000000,),dtype=tf.float32)
        else:
            nums = tf.range(200000000, dtype=tf.float32)
        i = tf.constant(0)
        def body(i, arr):
            arr = arr.write(i, func(nums))
            return i+1, arr
        i, arr = tf.while_loop(
            cond=lambda i, x: i < 1,
            body=body,
            loop_vars=[i, arr],
            parallel_iterations=1)
        dot_product = arr.read(0)
    with tf.Session(graph=graph) as sess:
        sess.run(dot_product)
        t1 = default_timer()
        for i in range(10):
            sess.run(dot_product)
        print("V2 timing: {} / run 10 runs".format((default_timer() - t1)/10.))

        profiler = tf.profiler.Profiler(sess.graph)
        run_meta = tf.RunMetadata()
        sess.run(dot_product,
                options = tf.RunOptions(trace_level = tf.RunOptions.FULL_TRACE),
                run_metadata = run_meta)
        profiler.add_step(0,run_meta)
        opts = (tf.profiler.ProfileOptionBuilder(
                 tf.profiler.ProfileOptionBuilder.time_and_memory())
                 .with_step(0)
                 .with_timeline_output('timeline_v2_{}'.format('random' if random else 'const')).build())
        profiler.profile_graph(options=opts)

if __name__ == '__main__':
    v1(False)
    v2(False)
    v1(True)
    v2(True)
```

**Comment by prb12** [comment]
OK, thanks - I don't see anything unusual in those results now other than v1 random being slower than v2 random.  I'd expect that running the RNG each step was substantially slower than the tf.range, but can't explain the -ve loop overhead unless it's just run-to-run variation. (this can happen based on other things running on the machine, heap allocation order, etc. etc.)

If the difference is repeatable, could you post the timelines and profile data?

**Comment by Joshuaalbert** [comment]
I'm able to repeat each time. 
V2 random is always a bit faster. 
Here are the timelines for the four situations.

[timelines.tar.gz](https://github.com/tensorflow/tensorflow/files/1729333/timelines.tar.gz)

**Comment by Joshuaalbert** [comment]
I think this should be closed, unless you still want to follow up that second case with random tensors in while loop being faster. Run MVCE to see if you get the same.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by apisutilis** [comment]
I'm going to close this, since the explanation above about the constant folder explains the initial behavior that I was seeing. Feel free to re-open if needed.
